### ISSUE TITLE
H5Store locking issues

### ISSUE THREAD

**Issue by bdice** [glotzerlab/signac#265]
### Description

I have been facing two issues that may or may not be related. Both issues are related to HDF5 file locking, a feature that was [implemented in HDF5 1.10.0 and changed in HDF5 1.10.1](https://support.hdfgroup.org/ftp/HDF5/releases/ReleaseFiles/hdf5-1.10.1-RELEASE.txt). This file locking feature is meant to help make SWMR (single writer, multiple reader) operations safer.

### To reproduce

The tests for storing pandas DataFrames in an H5Store fail. The file is already open from the `H5Store` instance, so when the `_h5get` function creates a [pandas `HDFStore` object](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#hdf5-pytables) (which uses pytables internally), the file fails to open because it's already locked by the `H5Store`.

### Error output

Example of a test failure message:
```python
======================================================================
ERROR: test_clear (__main__.H5StorePandasDataTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bdice/miniconda3/envs/py38/lib/python3.8/site-packages/pandas/io/pytables.py", line 627, in open
    self._handle = tables.open_file(self._path, self._mode, **kwargs)
  File "/home/bdice/miniconda3/envs/py38/lib/python3.8/site-packages/tables/file.py", line 315, in open_file
    return File(filename, mode, title, root_uep, filters, **kwargs)
  File "/home/bdice/miniconda3/envs/py38/lib/python3.8/site-packages/tables/file.py", line 778, in __init__
    self._g_new(filename, mode, **params)
  File "tables/hdf5extension.pyx", line 492, in tables.hdf5extension.File._g_new
tables.exceptions.HDF5ExtError: HDF5 error back trace

  File "H5F.c", line 509, in H5Fopen
    unable to open file
  File "H5Fint.c", line 1400, in H5F__open
    unable to open file
  File "H5Fint.c", line 1615, in H5F_open
    unable to lock the file
  File "H5FD.c", line 1640, in H5FD_lock
    driver lock request failed
  File "H5FDsec2.c", line 941, in H5FD_sec2_lock
    unable to lock file, errno = 11, error message = 'Resource temporarily unavailable'

End of HDF5 error back trace

Unable to open/create file '/tmp/signac_test_h5store_952j661f/signac_test_h5store.h5'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_h5store.py", line 286, in test_clear
    self.assertEqual(h5s[key], d)
  File "/home/bdice/code/signac/signac/core/h5store.py", line 405, in __getitem__
    return _h5get(self, self._file, key)
  File "/home/bdice/code/signac/signac/core/h5store.py", line 137, in _h5get
    with _pandas.HDFStore(grp.file.filename, mode='r') as store_:
  File "/home/bdice/miniconda3/envs/py38/lib/python3.8/site-packages/pandas/io/pytables.py", line 505, in __init__
    self.open(mode=mode, **kwargs)
  File "/home/bdice/miniconda3/envs/py38/lib/python3.8/site-packages/pandas/io/pytables.py", line 661, in open
    raise IOError(str(e))
OSError: HDF5 error back trace

  File "H5F.c", line 509, in H5Fopen
    unable to open file
  File "H5Fint.c", line 1400, in H5F__open
    unable to open file
  File "H5Fint.c", line 1615, in H5F_open
    unable to lock the file
  File "H5FD.c", line 1640, in H5FD_lock
    driver lock request failed
  File "H5FDsec2.c", line 941, in H5FD_sec2_lock
    unable to lock file, errno = 11, error message = 'Resource temporarily unavailable'

End of HDF5 error back trace

Unable to open/create file '/tmp/signac_test_h5store_952j661f/signac_test_h5store.h5'

----------------------------------------------------------------------
```

This configuration showed failures:
```bash
$ python -c "import tables; tables.print_versions(); import h5py; print(h5py.version.info)"
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
PyTables version:    3.6.1
HDF5 version:        1.10.4
NumPy version:       1.18.0
Numexpr version:     2.7.0 (not using Intel's VML/MKL)
Zlib version:        1.2.11 (in Python interpreter)
LZO version:         2.09 (Feb 04 2015)
BZIP2 version:       1.0.6 (6-Sept-2010)
Blosc version:       1.16.3 (2019-03-08)
Blosc compressors:   blosclz (1.1.0), lz4 (1.8.3), lz4hc (1.8.3), snappy (1.1.1), zlib (1.2.8), zstd (1.3.8)
Blosc filters:       shuffle, bitshuffle
Python version:      3.8.0 | packaged by conda-forge | (default, Nov  4 2019, 01:33:03)
[GCC 7.3.0]
Platform:            Linux-4.4.0-18362-Microsoft-x86_64-with-glibc2.10
Byte-ordering:       little
Detected cores:      4
Default encoding:    utf-8
Default FS encoding: utf-8
Default locale:      (en_US, UTF-8)
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Summary of the h5py configuration
---------------------------------

h5py    2.10.0
HDF5    1.10.4
Python  3.8.0 | packaged by conda-forge | (default, Nov  4 2019, 01:33:03)
[GCC 7.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.18.0
```

This configuration did **NOT** show failures on the same machine, so I suspect it's some combination of the version changes in this config. (I used two different conda environments.)

```bash
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
PyTables version:    3.6.1
HDF5 version:        1.10.5
NumPy version:       1.17.3
Numexpr version:     2.7.0 (not using Intel's VML/MKL)
Zlib version:        1.2.11 (in Python interpreter)
LZO version:         2.10 (Mar 01 2017)
BZIP2 version:       1.0.8 (13-Jul-2019)
Blosc version:       1.17.0 (2019-07-23)
Blosc compressors:   blosclz (2.0.0), lz4 (1.9.1), lz4hc (1.9.1), snappy (1.1.1), zlib (1.2.8), zstd (1.4.1)
Blosc filters:       shuffle, bitshuffle
Cython version:      0.29.13
Python version:      3.7.6 | packaged by conda-forge | (default, Dec 27 2019, 00:09:34)
[GCC 7.3.0]
Platform:            Linux-4.4.0-18362-Microsoft-x86_64-with-debian-buster-sid
Byte-ordering:       little
Detected cores:      4
Default encoding:    utf-8
Default FS encoding: utf-8
Default locale:      (en_US, UTF-8)
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Summary of the h5py configuration
---------------------------------

h5py    2.10.0
HDF5    1.10.5
Python  3.7.6 | packaged by conda-forge | (default, Dec 27 2019, 00:09:34)
[GCC 7.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.17.3
```

### System configuration

 - Operating System: Ubuntu 18.04
 - Version of Python: 3.7 or 3.8
 - Version of signac: 1.3

**Comment by vyasr** [comment]
@bdice is there anything we can do about this other than document the usage of `HDF5_USE_FILE_LOCKING` (with all appropriate "here there be dragons" warnings for SWMR users)?

**Comment by bdice** [comment]
@vyasr This was partly resolved by #266. Otherwise 
![image](https://user-images.githubusercontent.com/3943761/75376134-843e3300-589d-11ea-8bd0-0089736f5946.png)
